<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>EigenstratFormat · EigenstratFormat.jl</title><meta name="title" content="EigenstratFormat · EigenstratFormat.jl"/><meta property="og:title" content="EigenstratFormat · EigenstratFormat.jl"/><meta property="twitter:title" content="EigenstratFormat · EigenstratFormat.jl"/><meta name="description" content="Documentation for EigenstratFormat.jl."/><meta property="og:description" content="Documentation for EigenstratFormat.jl."/><meta property="twitter:description" content="Documentation for EigenstratFormat.jl."/><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="search_index.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>EigenstratFormat.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>EigenstratFormat</a><ul class="internal"><li><a class="tocitem" href="#Contents"><span>Contents</span></a></li><li><a class="tocitem" href="#Index"><span>Index</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>EigenstratFormat</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>EigenstratFormat</a></li></ul></nav><div class="docs-right"><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="EigenstratFormat"><a class="docs-heading-anchor" href="#EigenstratFormat">EigenstratFormat</a><a id="EigenstratFormat-1"></a><a class="docs-heading-anchor-permalink" href="#EigenstratFormat" title="Permalink"></a></h1><ul><li><a href="#EigenstratFormat">EigenstratFormat</a></li><li class="no-marker"><ul><li><a href="#Contents">Contents</a></li><li><a href="#Index">Index</a></li></ul></li></ul><h2 id="Contents"><a class="docs-heading-anchor" href="#Contents">Contents</a><a id="Contents-1"></a><a class="docs-heading-anchor-permalink" href="#Contents" title="Permalink"></a></h2><article><details class="docstring" open="true"><summary id="EigenstratFormat.EigenstratFormat"><a class="docstring-binding" href="#EigenstratFormat.EigenstratFormat"><code>EigenstratFormat.EigenstratFormat</code></a> — <span class="docstring-category">Module</span></summary><section><div><p>Methods for the Eigenstrat format that is commonly used in genetics.</p></div></section></details></article><article><details class="docstring" open="true"><summary id="EigenstratFormat._alleles-Tuple{Vector{UInt8}, Int64}"><a class="docstring-binding" href="#EigenstratFormat._alleles-Tuple{Vector{UInt8}, Int64}"><code>EigenstratFormat._alleles</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">_alleles(bytes::Vector{UInt8}, idx::Int64)</code></pre><p>Return the number of variant alleles for a specified individual.</p><p><code>bytes</code>: Row of bytes that encode an SNP for all individuals.</p><p><code>idx</code>: Index of the individual.</p></div></section></details></article><article><details class="docstring" open="true"><summary id="EigenstratFormat._bitpair-Tuple{UInt8, Integer}"><a class="docstring-binding" href="#EigenstratFormat._bitpair-Tuple{UInt8, Integer}"><code>EigenstratFormat._bitpair</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">_bitpair(byte::UInt8, pos::Int64)</code></pre><p>Extract 2 bits from a byte. The bit positions start at 0. Allowed are values 0, 1, 2, 3.</p></div></section></details></article><article><details class="docstring" open="true"><summary id="EigenstratFormat._encode-Tuple{Tuple{Char, Char}, UInt8, Int64, Tuple{Char, Char}}"><a class="docstring-binding" href="#EigenstratFormat._encode-Tuple{Tuple{Char, Char}, UInt8, Int64, Tuple{Char, Char}}"><code>EigenstratFormat._encode</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">_encode(
    genotype::Tuple{Char, Char},
    byte::UInt8,
    position::Int64,
    reference_snp::Tuple{Char, Char}
)</code></pre><p>Encode the given genotype in a byte at the given bitpair position (0..3).</p><p>Each SNP is characterized by a Tuple of two alleles. The <code>reference_snp</code> contains the reference allele and the derived allele.</p></div></section></details></article><article><details class="docstring" open="true"><summary id="EigenstratFormat._hashit-Tuple{AbstractString}"><a class="docstring-binding" href="#EigenstratFormat._hashit-Tuple{AbstractString}"><code>EigenstratFormat._hashit</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">_hashit(sequence::String)</code></pre><p>Calculate the hashsum of a String.</p><p>This is basically the same method as in Nick Patterson&#39;s original <a href="https://github.com/DReichLab/EIG/blob/master/src/admutils.c">C code</a>.</p><p>However, the original C version uses 32 bit integer values and integer overflows which may result in undefined behavior on some machines.  This method uses 64 bit integers to circumvent this problem.</p><p>This method will probably fail if the sequence contains non-ASCII Unicode characters. I do not know if this is defined in any way.</p></div></section></details></article><article><details class="docstring" open="true"><summary id="EigenstratFormat._hashsum-Tuple{Vector{&lt;:AbstractString}}"><a class="docstring-binding" href="#EigenstratFormat._hashsum-Tuple{Vector{&lt;:AbstractString}}"><code>EigenstratFormat._hashsum</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">_hashsum(sequences::Vector{&lt;:AbstractString})</code></pre><p>Calculate the hashsum of a Vector of Strings using the _hashit() method.</p><p>This is the same method as the <code>hasharr</code> function in Nick Patterson&#39;s <a href="https://github.com/DReichLab/EIG/blob/master/src/admutils.c">C code</a>.</p></div></section></details></article><article><details class="docstring" open="true"><summary id="EigenstratFormat._hashsum-Tuple{Vector{&lt;:AbstractString}}"><a class="docstring-binding" href="#EigenstratFormat._hashsum-Tuple{Vector{&lt;:AbstractString}}"><code>EigenstratFormat._hashsum</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">_hashsum(sequences::Vector{&lt;:AbstractString})</code></pre><p>Calculate the hashsum of a Vector of Strings using the _hashit() method.</p><p>Again this is basically the same method as in the software from David Reich&#39;s laboratory: https://github.com/DReichLab/EIG</p></div></section></details></article><article><details class="docstring" open="true"><summary id="EigenstratFormat.add_individual-NTuple{4, AbstractString}"><a class="docstring-binding" href="#EigenstratFormat.add_individual-NTuple{4, AbstractString}"><code>EigenstratFormat.add_individual</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">add_individual(
    inprefix::AbstractString,
    outprefix::AbstractString,
    ind_snp_file::AbstractString,
    id::AbstractString;
    gender = &quot;U&quot;,
    status = &quot;Control&quot;
)</code></pre><p>Add an individual to a database in Eigenstrat format. </p><p>The SNPs in the database remain untouched. If the individual displayes SNPs that are not listed in the database or multiallelic ones those SNPs are removed.</p><p><code>inprefix</code>: Prefix of the input database.</p><p><code>outprefix</code>: Prefix of the output database.</p><p><code>ind_snp_file</code>: File containing SNP results for the individual. This should     work with files from Family Tree DNA Family Finder, MyHeritage, LivingDNA     and 23andMe.</p><p><code>id</code>: ID of the individual. For living persons I recommed the name.</p><p><code>gender</code>: U, F or M (Unknown, Female or Male)</p><p><code>status</code>: Control, Case or a population label.</p></div></section></details></article><article><details class="docstring" open="true"><summary id="EigenstratFormat.hash_ids-Tuple{AbstractString}"><a class="docstring-binding" href="#EigenstratFormat.hash_ids-Tuple{AbstractString}"><code>EigenstratFormat.hash_ids</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">hash_ids(filename::AbstractString)</code></pre><p>Create the hashsum of .snp and .ind files. SNP and Individual files contain an ID in the first column. This method uses those IDs to calculate a hashsum. The sums are needed to store genotype data in packed format.</p></div></section></details></article><article><details class="docstring" open="true"><summary id="EigenstratFormat.read_eigenstrat_geno-Tuple{AbstractString, Int64, Int64}"><a class="docstring-binding" href="#EigenstratFormat.read_eigenstrat_geno-Tuple{AbstractString, Int64, Int64}"><code>EigenstratFormat.read_eigenstrat_geno</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">read_eigenstrat_geno(
    genofile::AbstractString,
    nsnp::Int64,
    nind::Int64;
    ind_idx::Vector{Int64} = [i for i = 1:nind]
)</code></pre><p>Read a genofile in PackedAncestryMap format. The file must be unzipped.</p><p><code>genofile</code>: filename</p><p><code>nsnp</code>: number of SNPs listed in .snp file.</p><p><code>nind</code>: number of individuals listed in .ind file.</p><p><code>ind_idx</code>: Indices of individuals that should be read from the file.</p><p>XXX Check for comment lines in .snp and .ind files.</p><p>File description: File header starts with GENO or TGENO (transposed GENO). So far files in the AADR archive seem to be GENO. So this method does not support the transposed TGENO format.</p><p>The text format contains one line per genotype:</p><p>SNP<em>ID  Sample</em>ID   Number<em>of</em>variant_alleles</p><p>The packed format:</p><p>Each SNP entry has 2 bits: 0, 1, 2, 3=missing, that denote the number of variant alleles as described at <a href="https://reich.hms.harvard.edu/software/InputFileFormats">David Reich&#39;s laboratory</a>.</p></div></section></details></article><article><details class="docstring" open="true"><summary id="EigenstratFormat.read_eigenstrat_ind-Tuple{AbstractString}"><a class="docstring-binding" href="#EigenstratFormat.read_eigenstrat_ind-Tuple{AbstractString}"><code>EigenstratFormat.read_eigenstrat_ind</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">read_eigenstrat_ind(indfile::AbstractString)</code></pre><p>Read individuals from Eigenstrat .ind file. The IND flle contains information about each individual in the database.</p><p>Return a <code>DataFrame</code> consisting of the columns:</p><p>ID, Gender, Status</p><p>where</p><p>Gender: M (male), F (Female) or U (unknown).</p><p>Status: Case, Control or population label.</p></div></section></details></article><article><details class="docstring" open="true"><summary id="EigenstratFormat.read_eigenstrat_snp-Tuple{AbstractString}"><a class="docstring-binding" href="#EigenstratFormat.read_eigenstrat_snp-Tuple{AbstractString}"><code>EigenstratFormat.read_eigenstrat_snp</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">read_eigenstrat_snp(snpfile::AbstractString)</code></pre><p>Read Eigenstrat .snp file. The SNP file contains information about each SNP.</p><p>Return a <code>DataFrame</code> containing the columns:</p><p>chromosome, rsid, cM, position, allele1, allele2</p></div></section></details></article><article><details class="docstring" open="true"><summary id="EigenstratFormat.read_snp_file-Tuple{AbstractString}"><a class="docstring-binding" href="#EigenstratFormat.read_snp_file-Tuple{AbstractString}"><code>EigenstratFormat.read_snp_file</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">read_snp_file(filename::AbstractString)</code></pre><p>Read file with autosomal results from FTDNA Family Finder, My Heritage or LivingDNA. Should also work with 24andMe files but not tested.</p><p>Return a <code>DataFrame</code> containing the columns:</p><p>rsid  chromosome  position  genotype</p></div></section></details></article><article><details class="docstring" open="true"><summary id="EigenstratFormat.write_23andMe-Tuple{AbstractString, Any}"><a class="docstring-binding" href="#EigenstratFormat.write_23andMe-Tuple{AbstractString, Any}"><code>EigenstratFormat.write_23andMe</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">write_23andMe(filename::AbstractString, snptable)</code></pre><p>Write a table of SNPs in 23andMe file format.</p><p>This method is included to satisfy users who use <a href="https://www.cog-genomics.org/plink2/">Plink</a>. Plink supports 23andMe files.</p><p>The snp table must satisfy the <a href="https://github.com/JuliaData/Tables.jl">Tables.jl interface</a>.</p><p>The table must contain the columns:</p><p>rsid, chromosome, position, genotype</p><p>However exact spelling is not mandatory.</p></div></section></details></article><article><details class="docstring" open="true"><summary id="EigenstratFormat.write_eigenstrat_geno-Tuple{AbstractString, Matrix{UInt8}}"><a class="docstring-binding" href="#EigenstratFormat.write_eigenstrat_geno-Tuple{AbstractString, Matrix{UInt8}}"><code>EigenstratFormat.write_eigenstrat_geno</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">write_eigenstrat_geno(
    genofile::AbstractString,
    genomatrix::Matrix{UInt8};
    ind_hash::Int64 = 0,
    snp_hash::Int64 = 0
)</code></pre><p>Write a genotype matrix to file in PackedAncestryMap format.</p><p><code>ind_hash</code>: Hashsum of .ind file.</p><p><code>snp_hash</code>: Hashsum of .snp file.</p></div></section></details></article><article><details class="docstring" open="true"><summary id="EigenstratFormat.write_eigenstrat_ind-Tuple{AbstractString, DataFrames.AbstractDataFrame}"><a class="docstring-binding" href="#EigenstratFormat.write_eigenstrat_ind-Tuple{AbstractString, DataFrames.AbstractDataFrame}"><code>EigenstratFormat.write_eigenstrat_ind</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">write_eigenstrat_ind(filename::AbstractString, inds::AbstractDataFrame)</code></pre><p>Write information about each individual to an .ind file.</p><p>The parameter <code>inds</code> contains information about each individual.</p><p>The <code>DataFrame</code> must have the columns</p><p>ID, Gender, Status</p></div></section></details></article><article><details class="docstring" open="true"><summary id="EigenstratFormat.write_eigenstrat_snp-Tuple{AbstractString, DataFrames.AbstractDataFrame}"><a class="docstring-binding" href="#EigenstratFormat.write_eigenstrat_snp-Tuple{AbstractString, DataFrames.AbstractDataFrame}"><code>EigenstratFormat.write_eigenstrat_snp</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">write_eigenstrat_snp(filename::AbstractString, snps::DataFrame)</code></pre><p>Write .snp file in Eigenstrat format. The SNP file contains information about each SNP. The SNPs are provided as a <code>DataFrame</code> in parameter <code>snps</code>.</p><p>The DataFrame must consist of the following columns:</p><p>chromosome, rsid, cM, position, allele1, allele2</p></div></section></details></article><h2 id="Index"><a class="docs-heading-anchor" href="#Index">Index</a><a id="Index-1"></a><a class="docs-heading-anchor-permalink" href="#Index" title="Permalink"></a></h2><ul><li><a href="#EigenstratFormat.EigenstratFormat"><code>EigenstratFormat.EigenstratFormat</code></a></li><li><a href="#EigenstratFormat._alleles-Tuple{Vector{UInt8}, Int64}"><code>EigenstratFormat._alleles</code></a></li><li><a href="#EigenstratFormat._bitpair-Tuple{UInt8, Integer}"><code>EigenstratFormat._bitpair</code></a></li><li><a href="#EigenstratFormat._encode-Tuple{Tuple{Char, Char}, UInt8, Int64, Tuple{Char, Char}}"><code>EigenstratFormat._encode</code></a></li><li><a href="#EigenstratFormat._hashit-Tuple{AbstractString}"><code>EigenstratFormat._hashit</code></a></li><li><a href="#EigenstratFormat._hashsum-Tuple{Vector{&lt;:AbstractString}}"><code>EigenstratFormat._hashsum</code></a></li><li><a href="#EigenstratFormat._hashsum-Tuple{Vector{&lt;:AbstractString}}"><code>EigenstratFormat._hashsum</code></a></li><li><a href="#EigenstratFormat.add_individual-NTuple{4, AbstractString}"><code>EigenstratFormat.add_individual</code></a></li><li><a href="#EigenstratFormat.hash_ids-Tuple{AbstractString}"><code>EigenstratFormat.hash_ids</code></a></li><li><a href="#EigenstratFormat.read_eigenstrat_geno-Tuple{AbstractString, Int64, Int64}"><code>EigenstratFormat.read_eigenstrat_geno</code></a></li><li><a href="#EigenstratFormat.read_eigenstrat_ind-Tuple{AbstractString}"><code>EigenstratFormat.read_eigenstrat_ind</code></a></li><li><a href="#EigenstratFormat.read_eigenstrat_snp-Tuple{AbstractString}"><code>EigenstratFormat.read_eigenstrat_snp</code></a></li><li><a href="#EigenstratFormat.read_snp_file-Tuple{AbstractString}"><code>EigenstratFormat.read_snp_file</code></a></li><li><a href="#EigenstratFormat.write_23andMe-Tuple{AbstractString, Any}"><code>EigenstratFormat.write_23andMe</code></a></li><li><a href="#EigenstratFormat.write_eigenstrat_geno-Tuple{AbstractString, Matrix{UInt8}}"><code>EigenstratFormat.write_eigenstrat_geno</code></a></li><li><a href="#EigenstratFormat.write_eigenstrat_ind-Tuple{AbstractString, DataFrames.AbstractDataFrame}"><code>EigenstratFormat.write_eigenstrat_ind</code></a></li><li><a href="#EigenstratFormat.write_eigenstrat_snp-Tuple{AbstractString, DataFrames.AbstractDataFrame}"><code>EigenstratFormat.write_eigenstrat_snp</code></a></li></ul></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.16.1 on <span class="colophon-date" title="Wednesday 18 February 2026 09:51">Wednesday 18 February 2026</span>. Using Julia version 1.12.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
